File: lib/models/captured_exchange.dart
import 'dart:typed_data';

class HttpHeader {
  final String name;
  final String value;
  const HttpHeader(this.name, this.value);
}

enum ExchangeState { inProgress, completed, failed }

class CapturedExchange {
  final String id;
  final DateTime startTime;
  DateTime? endTime;
  final String method;
  final String url;
  final String scheme;
  final String host;
  final int port;
  final String path;
  final List<HttpHeader> requestHeaders;
  final String? requestBodyRef;
  final int requestSize;
  int? statusCode;
  String? statusMessage;
  List<HttpHeader>? responseHeaders;
  String? responseBodyRef;
  int? responseSize;
  final bool isHTTPS;
  final bool isMITMDecrypted;
  ExchangeState state;
  String? errorMessage;

  // Lazily fetched body bytes (cached after first fetch)
  Uint8List? _cachedRequestBody;
  Uint8List? _cachedResponseBody;

  Uint8List? get cachedRequestBody => _cachedRequestBody;
  Uint8List? get cachedResponseBody => _cachedResponseBody;
  void setCachedRequestBody(Uint8List data) => _cachedRequestBody = data;
  void setCachedResponseBody(Uint8List data) => _cachedResponseBody = data;

  Duration? get duration =>
      endTime != null ? endTime!.difference(startTime) : null;

  CapturedExchange({
    required this.id,
    required this.startTime,
    this.endTime,
    required this.method,
    required this.url,
    required this.scheme,
    required this.host,
    required this.port,
    required this.path,
    required this.requestHeaders,
    this.requestBodyRef,
    required this.requestSize,
    this.statusCode,
    this.statusMessage,
    this.responseHeaders,
    this.responseBodyRef,
    this.responseSize,
    required this.isHTTPS,
    required this.isMITMDecrypted,
    this.state = ExchangeState.inProgress,
    this.errorMessage,
  });

  factory CapturedExchange.fromMap(Map<Object?, Object?> raw) {
    final map = Map<String, dynamic>.from(raw);
    return CapturedExchange(
      id: map['id'] as String,
      startTime: DateTime.fromMillisecondsSinceEpoch(
          (map['startTime'] as num).toInt()),
      endTime: map['endTime'] != null
          ? DateTime.fromMillisecondsSinceEpoch(
              (map['endTime'] as num).toInt())
          : null,
      method: map['method'] as String,
      url: map['url'] as String,
      scheme: map['scheme'] as String,
      host: map['host'] as String,
      port: map['port'] as int,
      path: map['path'] as String,
      requestHeaders: _parseHeaders(map['requestHeaders']),
      requestBodyRef: map['requestBodyRef'] as String?,
      requestSize: map['requestSize'] as int,
      statusCode: map['statusCode'] as int?,
      statusMessage: map['statusMessage'] as String?,
      responseHeaders: map['responseHeaders'] != null
          ? _parseHeaders(map['responseHeaders'])
          : null,
      responseBodyRef: map['responseBodyRef'] as String?,
      responseSize: map['responseSize'] as int?,
      isHTTPS: map['isHTTPS'] as bool,
      isMITMDecrypted: map['isMITMDecrypted'] as bool,
      state: _parseState(map['state'] as String),
      errorMessage: map['errorMessage'] as String?,
    );
  }

  /// Merges updated fields from a channel "update" event into this exchange.
  void applyUpdate(CapturedExchange updated) {
    endTime = updated.endTime;
    statusCode = updated.statusCode;
    statusMessage = updated.statusMessage;
    responseHeaders = updated.responseHeaders;
    responseSize = updated.responseSize;
    state = updated.state;
    errorMessage = updated.errorMessage;
    // Don't overwrite cached bodies — keep existing refs only if null
    if (_cachedResponseBody == null && updated.responseBodyRef != null) {
      // ref updated — invalidate old cached body
      _cachedResponseBody = null;
    }
  }

  static List<HttpHeader> _parseHeaders(dynamic raw) {
    if (raw == null) return [];
    final list = raw as List<dynamic>;
    return list.map((e) {
      final m = Map<String, dynamic>.from(e as Map);
      return HttpHeader(m['name'] as String, m['value'] as String);
    }).toList();
  }

  static ExchangeState _parseState(String s) => switch (s) {
        'completed' => ExchangeState.completed,
        'failed' => ExchangeState.failed,
        _ => ExchangeState.inProgress,
      };
}

File: packages/rox_proxy_native/macos/rox_proxy_native/Sources/rox_proxy_native/Models/CapturedExchange.swift
import Foundation

struct CapturedExchange: Identifiable, Sendable {
    let id: UUID
    let startTime: Date
    var endTime: Date?

    // Request
    var method: String
    var url: String
    var scheme: String          // "http" or "https"
    var host: String
    var port: Int               // target port (80, 443, or custom)
    var path: String            // path + query, parsed from URL
    var requestHeaders: [(name: String, value: String)]
    var requestBody: BodyContent?

    // Response (nil while in-flight)
    var statusCode: Int?
    var statusMessage: String?
    var responseHeaders: [(name: String, value: String)]?
    var responseBody: BodyContent?

    // Metadata
    var duration: TimeInterval? { endTime.map { $0.timeIntervalSince(startTime) } }
    var requestSize: Int
    var responseSize: Int?
    var isHTTPS: Bool
    var isMITMDecrypted: Bool
    var state: ExchangeState

    enum ExchangeState: Sendable, Equatable {
        case inProgress
        case completed
        case failed(String)
    }

    init(
        id: UUID = UUID(),
        startTime: Date = Date(),
        method: String,
        url: String,
        scheme: String = "http",
        host: String,
        port: Int = 80,
        requestHeaders: [(name: String, value: String)] = [],
        requestBody: BodyContent? = nil,
        requestSize: Int = 0,
        isHTTPS: Bool = false,
        isMITMDecrypted: Bool = false
    ) {
        self.id = id
        self.startTime = startTime
        self.method = method
        self.url = url
        self.scheme = scheme
        self.host = host
        self.port = port
        self.requestHeaders = requestHeaders
        self.requestBody = requestBody
        self.requestSize = requestSize
        self.isHTTPS = isHTTPS
        self.isMITMDecrypted = isMITMDecrypted
        self.state = .inProgress

        // Parse path from URL (requires a valid absolute URL with host), fallback to "/"
        if let parsed = URL(string: url), parsed.host != nil {
            let pathStr = parsed.path.isEmpty ? "/" : parsed.path
            var full = pathStr
            if let query = parsed.query { full += "?" + query }
            self.path = full
        } else {
            self.path = "/"
        }
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

// MARK: - BodyContent

enum BodyContent: Sendable, Equatable {
    case data(Data)
    case truncated(Data, totalSize: Int)
    case empty

    static let maxInMemorySize = 10 * 1024 * 1024 // 10 MB

    // MARK: Convenience

    var isEmpty: Bool {
        if case .empty = self { return true }
        return false
    }

    /// The bytes stored in memory (may be partial for .truncated).
    var data: Data? {
        switch self {
        case .data(let d): return d
        case .truncated(let d, _): return d
        case .empty: return nil
        }
    }

    var isTruncated: Bool {
        if case .truncated = self { return true }
        return false
    }

    /// The actual total body size (data.count for .data, reported totalSize for .truncated).
    var totalSize: Int {
        switch self {
        case .data(let d): return d.count
        case .truncated(_, let size): return size
        case .empty: return 0
        }
    }

    /// Attempts to decode the stored bytes as a UTF-8 string (or the given encoding).
    func asString(encoding: String.Encoding = .utf8) -> String? {
        guard let bytes = data, !bytes.isEmpty else { return nil }
        return String(data: bytes, encoding: encoding)
    }

    // MARK: Builder

    /// Appends raw bytes to the body, respecting the in-memory cap.
    /// Call this repeatedly as body chunks arrive from the network.
    static func appending(existing: BodyContent?, newBytes: Data, runningTotal: inout Int) -> BodyContent {
        runningTotal += newBytes.count

        switch existing {
        case nil, .empty:
            if newBytes.count > maxInMemorySize {
                return .truncated(newBytes.prefix(maxInMemorySize), totalSize: runningTotal)
            }
            return .data(newBytes)

        case .data(var accumulated):
            accumulated.append(newBytes)
            if accumulated.count > maxInMemorySize {
                return .truncated(accumulated.prefix(maxInMemorySize), totalSize: runningTotal)
            }
            return .data(accumulated)

        case .truncated(let partial, _):
            // Already truncated — update total size only
            return .truncated(partial, totalSize: runningTotal)

        case .some(_):
            return existing ?? .empty
        }
    }
}
